### PR TITLE
Api/mastodon

### PR DESCRIPTION
**Pull request by benetherington**

### CODE REVIEW

File: title-suggestions/slash/utilities/ai-utilities.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        size: '256x256',
+    });
+    const resp = await fetch(url, {method, headers, body});
+    const jsn = await resp.json();
+    console.log(JSON.stringify(jsn));
+    return jsn;
+    // 'These Engines Go to 11' https://oaidalleapiprodscus.blob.core.windows.net/private/org-sBbtuabVZwBpVB0Oo7ex1emP/user-yaeCG5uOkrlxb56hqukzfkjV/img-3ZfNOI5FQwGvA0CdoyNlwrv1.png?st=2022-12-05T16%3A03%3A30Z&se=2022-12-05T18%3A03%3A30Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2022-12-05T12%3A15%3A20Z&ske=2022-12-06T12%3A15%3A20Z&sks=b&skv=2021-08-06&sig=gKNTKdnBsXPGeoJJ71w4TLNlj%2BnSUv5dfgOwL28PBEc%3D
+    // 'NASA Cuts Carbs' https://oaidalleapiprodscus.blob.core.windows.net/private/org-sBbtuabVZwBpVB0Oo7ex1emP/user-yaeCG5uOkrlxb56hqukzfkjV/img-LN9za2s9QCDTjwakuE690f6b.png?st=2022-12-05T16%3A00%3A53Z&se=2022-12-05T18%3A00%3A53Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2022-12-05T12%3A48%3A31Z&ske=2022-12-06T12%3A48%3A31Z&sks=b&skv=2021-08-06&sig=h2Cr9mVeS4YEROFsS5hms5Pt%2Bz70d7HDewIJS91a7GQ%3D
```
Comment (fonji) [inline title-suggestions/slash/utilities/ai-utilities.js:24]: what are those?

File: twsf/mastodon/toots.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+const fetch = require('node-fetch');
+const logger = require('../../logger');
+const {addNewGuess, guessTypes} = require('../../database/twsf');
+
+// const headers = {
+//     Authorization: `Bearer ${process.env.MASTODON_TOKEN}`,
+// };
+
+/*
+const fetchTwsfTootHistory = async (instance) => {
```
Comment (fonji) [inline twsf/mastodon/toots.js:12]: useful or :broom: ?

File: twsf/mastodon/toots.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        );
+        return [];
+    }
+
+    const jsn = await resp.json();
+    return jsn.hashtags.find((ht) => ht.name === 'thisweeksf').history;
+};
+*/
+
+const tootIsRemote = (status) => {
```
Comment (fonji) [inline twsf/mastodon/toots.js:32]: by the way: why `const` and not function?

File: twsf/mastodon/toots.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    status.account.acct.includes('@');
+};
+const getTootInstanceName = (status) => {
```
Comment (fonji) [inline twsf/mastodon/toots.js:35]: add a whiteline between them?

File: twsf/mastodon/toots.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    return status.url.match(/https?:\/\/(.*?)\//)[1];
+};
+const normalizeAcct = (status) => {
```
Comment (fonji) [inline twsf/mastodon/toots.js:38]: `account` is too many letters? :trollface:

File: twsf/mastodon/toots.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    const replies = jsn.descendants;
+
+    const selfReplies = replies.filter(
+        (reply) =>
+            reply.account.id === status.account.id &&
+            reply.in_reply_to_account_id === status.account.id,
+    );
+    const replyTexts = Promise.all(selfReplies.map(getFullText));
+
+    return replyTexts;
```
Comment (fonji) [inline twsf/mastodon/toots.js:98]: no need for a variable, you can return the `Promise.all` call, no?

File: twsf/mastodon/toots.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    const instance = 'http://' + getTootInstanceName(status);
+    const path = `/api/v1/statuses/${status.id}/context`;
+    const url = instance + path;
+
+    const resp = await fetch(url);
+    if (!resp.ok) {
+        console.error(
+            `Encountered ${resp.statusText} when fetching replies to ${status.id}`,
+        );
+        return [];
```
Comment (fonji) [inline twsf/mastodon/toots.js:85]: as this is not a Promise, you may get an error in the caller if it ever happens
(I think async / await may be really nice to manage these kind of things so I may be wrong)

**Review by fonji — COMMENTED** [review]
I'm passing by _real quick_, I didn't read everything.
Also don't worry about ignoring my comments :smile:

**Comment by benetherington** [comment]
Ah, beautiful, thank you! Your eye is invaluable.

**Comment by fonji** [comment]
oh stop it you

File: database/migrations/public/006-add-twsf-toot.sql [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    displayName         TEXT,
+    
+    twitterId           TEXT UNIQUE,
+    twitterUsername     TEXT UNIQUE,
+    twitterDisplayName  TEXT,
+    
+    emailAddress        TEXT UNIQUE,
+    emailName           TEXT,
+    
+    mastodonId          TEXT,
```
[the remark this answers is not in the document]
Comment (benetherington) [reply database/migrations/public/006-add-twsf-toot.sql:76]: This is actually just the last part of the URI. It's scoped to an instance, so there could be a collision here, but Mastodon tries to make them unique across instances. 

For reference, the toots response includes:
```json
{
    id: '110610931685237010',
    uri: 'https://spacey.space/users/benetherington/statuses/110610931685237010',
    url: 'https://spacey.space/@benetherington/110610931685237010'
}
```

File: src/views/layout.pug [hunk 9]
```diff
@@ -40,3 +40,6 @@ html
                 
         div#content
             block content
+
+        div#templates(hidden)
+            block templates
```
[the remark this answers is not in the document]
Comment (benetherington) [reply src/views/layout.pug:45]: I S2G it's set up that way, but after glancing at this a month ago, I've noticed a bunch of other projects also have an issue with this. At some point I'll track down the RC file that's wrecking havoc...

File: src/views/twsf.js [hunk 10]
```diff
@@ -76,47 +120,71 @@ const createGuessRow = (guess) => {
...
+    rowContainer.querySelector('#correct-id').id = `correct-${guess.guessId}`;
+    rowContainer.querySelector(
+        '[for=correct-id]',
+    ).for = `correct-${guess.guessId}`;
+
+    rowContainer.querySelector('#bonus-id').name = `points-${guess.guessId}`;
+    rowContainer.querySelector('#bonus-id').id = `bonus-${guess.guessId}`;
+    rowContainer.querySelector('[for=bonus-id]').for = `bonus-${guess.guessId}`;
+
+    // rowContainer.innerHTML = `
```
[the remark this answers is not in the document]
Comment (benetherington) [reply src/views/twsf.js:153]: Code fast, break things.

File: database/migrations/public/006-add-twsf-toot.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (fonji) [reply database/migrations/public/006-add-twsf-toot.sql:76]: Interesting, thanks!

File: src/views/twsf.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (fonji) [reply src/views/twsf.js:153]: Seems like a recipe for happiness

**Pull request merged**